File: ActionUINodeJS/test_insert_api.js
'use strict';
/**
 * test_insert_api.js — Runtime structural mutations smoke tests.
 *
 * Exercises insertElement, insertRow, removeElement at both the native (N-API)
 * and JS wrapper (Window class) layers without a running NSApplication.
 * No UI is displayed — tests verify API surface, type correctness, and that
 * unknown UUIDs produce the expected errors without crashing.
 *
 * Run: node test_insert_api.js
 */

const actionui = require('./index.js');
const native = actionui._native;

let failures = [];

function check(label, condition) {
    if (condition) {
        console.log(`  [PASS] ${label}`);
    } else {
        console.error(`  [FAIL] ${label}`);
        failures.push(label);
    }
    return condition;
}

const FAKE_UUID          = '00000000-0000-0000-0000-000000000000';
const VALID_ELEMENT_JSON = '{"type":"Text","properties":{"text":"hello"}}';
const VALID_CELLS_JSON   = '[{"type":"Text","properties":{"text":"C0"}},{"type":"Text","properties":{"text":"C1"}}]';

// ---------------------------------------------------------------------------
// Native API surface
// ---------------------------------------------------------------------------
function testNativeAPISurface() {
    console.log('\n=== Native module: structural mutation API surface ===');

    for (const name of ['insertElement', 'insertRow', 'removeElement']) {
        check(`native.${name} is present`, typeof native[name] === 'function');
    }
}

// ---------------------------------------------------------------------------
// Window method surface
// ---------------------------------------------------------------------------
function testWindowMethodSurface(win) {
    console.log('\n=== Window: structural mutation method surface ===');

    for (const name of ['insertElement', 'insertRow', 'removeElement']) {
        check(`win.${name} is a function`, typeof win[name] === 'function');
    }
}

// ---------------------------------------------------------------------------
// insertElement — unknown UUID throws
// ---------------------------------------------------------------------------
function testInsertElementUnknownUUID(win) {
    console.log('\n=== insertElement: unknown UUID throws ===');

    // Native layer
    let threw = false;
    try { native.insertElement(FAKE_UUID, 1n, VALID_ELEMENT_JSON, null, actionui.InsertPosition.APPEND, 0n); }
    catch { threw = true; }
    check('native.insertElement(unknown UUID) throws', threw);

    // JS wrapper (calls native internally)
    threw = false;
    try { win.insertElement(1, VALID_ELEMENT_JSON); }
    catch { threw = true; }
    check('win.insertElement(unknown UUID) throws', threw);
}

// ---------------------------------------------------------------------------
// insertElement — object input coerced to JSON
// ---------------------------------------------------------------------------
function testInsertElementObjectInput(win) {
    console.log('\n=== insertElement: object input coerced to JSON ===');

    const element = { type: 'Text', properties: { text: 'hello' } };
    let threw = false;
    try { win.insertElement(1, element); }
    catch (e) {
        // Should throw because UUID is unknown, not because of a TypeError
        threw = !(e instanceof TypeError);
    }
    check('win.insertElement(object) coerced to JSON (throws unknown-UUID, not TypeError)', threw);
}

// ---------------------------------------------------------------------------
// insertRow — unknown UUID throws
// ---------------------------------------------------------------------------
function testInsertRowUnknownUUID(win) {
    console.log('\n=== insertRow: unknown UUID throws ===');

    // Native layer
    let threw = false;
    try { native.insertRow(FAKE_UUID, 2n, VALID_CELLS_JSON, null, 0n, 0n); }
    catch { threw = true; }
    check('native.insertRow(unknown UUID) throws', threw);

    // JS wrapper
    threw = false;
    try { win.insertRow(2, VALID_CELLS_JSON); }
    catch { threw = true; }
    check('win.insertRow(unknown UUID) throws', threw);
}

// ---------------------------------------------------------------------------
// insertRow — array input coerced to JSON
// ---------------------------------------------------------------------------
function testInsertRowArrayInput(win) {
    console.log('\n=== insertRow: array input coerced to JSON ===');

    const cells = [
        { type: 'Text', properties: { text: 'C0' } },
        { type: 'Text', properties: { text: 'C1' } },
    ];
    let threw = false;
    try { win.insertRow(2, cells); }
    catch (e) {
        threw = !(e instanceof TypeError);
    }
    check('win.insertRow(array) coerced to JSON (throws unknown-UUID, not TypeError)', threw);
}

// ---------------------------------------------------------------------------
// removeElement — unknown UUID throws
// ---------------------------------------------------------------------------
function testRemoveElementUnknownUUID(win) {
    console.log('\n=== removeElement: unknown UUID throws ===');

    // Native layer
    let threw = false;
    try { native.removeElement(FAKE_UUID, 10n); }
    catch { threw = true; }
    check('native.removeElement(unknown UUID) throws', threw);

    // JS wrapper
    threw = false;
    try { win.removeElement(10); }
    catch { threw = true; }
    check('win.removeElement(unknown UUID) throws', threw);
}

// ---------------------------------------------------------------------------
// Position / optional arg defaults
// ---------------------------------------------------------------------------
function testPositionDefaults(win) {
    console.log('\n=== insertElement: optional args default correctly ===');

    const cases = [
        ['no container/position',         []],
        ['container=null',                [null]],
        ['container="children"',          ['children']],
        ['position=PREPEND',         [null, actionui.InsertPosition.PREPEND]],
        ['position=AT, param=0',     [null, actionui.InsertPosition.AT, 0]],
        ['position=BEFORE, param=9', [null, actionui.InsertPosition.BEFORE, 9]],
        ['position=AFTER, param=9',  [null, actionui.InsertPosition.AFTER, 9]],
    ];

    for (const [desc, extraArgs] of cases) {
        let threw = false, typeError = false;
        try { win.insertElement(1, VALID_ELEMENT_JSON, ...extraArgs); }
        catch (e) {
            threw = true;
            typeError = e instanceof TypeError;
        }
        check(`insertElement(${desc}) throws unknown-UUID error, not TypeError`,
              threw && !typeError);
    }
}

function testInsertRowPositionDefaults(win) {
    console.log('\n=== insertRow: optional args default correctly ===');

    const cases = [
        ['no container/position',       []],
        ['position=PREPEND',   [null, actionui.InsertPosition.PREPEND]],
        ['position=AT, index=0',    [null, actionui.InsertPosition.AT, 0]],
    ];

    for (const [desc, extraArgs] of cases) {
        let threw = false, typeError = false;
        try { win.insertRow(2, VALID_CELLS_JSON, ...extraArgs); }
        catch (e) {
            threw = true;
            typeError = e instanceof TypeError;
        }
        check(`insertRow(${desc}) throws unknown-UUID error, not TypeError`,
              threw && !typeError);
    }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
console.log('ActionUI Structural Mutation API Smoke Tests');
console.log('='.repeat(55));
console.log('(No NSApplication run loop — tests the N-API binding layer only)');

const app = new actionui.Application({ name: 'InsertTestApp' });
const win = new actionui.Window(FAKE_UUID);

testNativeAPISurface();
testWindowMethodSurface(win);
testInsertElementUnknownUUID(win);
testInsertElementObjectInput(win);
testInsertRowUnknownUUID(win);
testInsertRowArrayInput(win);
testRemoveElementUnknownUUID(win);
testPositionDefaults(win);
testInsertRowPositionDefaults(win);

console.log('\n' + '='.repeat(55));
if (failures.length === 0) {
    console.log('All checks PASSED.');
    process.exit(0);
} else {
    console.error(`FAILED — ${failures.length} check(s):`);
    for (const f of failures) console.error(`  - ${f}`);
    process.exit(1);
}

File: ActionUIPython/test_insert_api.py
#!/usr/bin/env python3
"""
Smoke tests for the runtime structural mutations Python bridge.

Exercises insert_element, insert_row, remove_element at both the C extension
(_actionui) and high-level (actionui.Window) layers without a live NSApplication
run loop. No UI is displayed — these tests verify API surface, type correctness,
and that unknown UUIDs produce the expected errors without crashing.

Run: python3 test_insert_api.py
"""

import sys
import json
import inspect
import _actionui
import actionui

# ---------------------------------------------------------------------------
# Simple pass/fail harness
# ---------------------------------------------------------------------------

_failures = []

def check(label: str, condition: bool) -> bool:
    status = "PASS" if condition else "FAIL"
    print(f"  [{status}] {label}")
    if not condition:
        _failures.append(label)
    return condition


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

_FAKE_UUID = "00000000-0000-0000-0000-000000000000"
_VALID_ELEMENT_JSON = '{"type":"Text","properties":{"text":"hello"}}'
_VALID_CELLS_JSON   = '[{"type":"Text","properties":{"text":"C0"}},{"type":"Text","properties":{"text":"C1"}}]'


def test_c_api_surface():
    """All three mutation functions must be present and callable in _actionui."""
    print("\n=== _actionui: structural mutation API surface ===")

    for name in ("insert_element", "insert_row", "remove_element"):
        check(f"_actionui.{name} is callable",
              callable(getattr(_actionui, name, None)))


def test_window_method_surface():
    """Window must expose insert_element, insert_row, remove_element."""
    print("\n=== Window: structural mutation method surface ===")

    W = actionui.Window
    for name in ("insert_element", "insert_row", "remove_element"):
        check(f"Window has method '{name}'",
              callable(getattr(W, name, None)))


def test_insert_element_signatures():
    """insert_element / insert_row / remove_element must have the expected parameters."""
    print("\n=== Window: method signatures ===")

    W = actionui.Window

    sig = inspect.signature(W.insert_element)
    params = list(sig.parameters.keys())
    check("insert_element has 'parent_id'",       "parent_id" in params)
    check("insert_element has 'element'",         "element" in params)
    check("insert_element has 'container'",            "container" in params)
    check("insert_element has 'position'",        "position" in params)
    check("insert_element has 'position_param'",  "position_param" in params)

    sig = inspect.signature(W.insert_row)
    params = list(sig.parameters.keys())
    check("insert_row has 'parent_id'",        "parent_id" in params)
    check("insert_row has 'cells'",            "cells" in params)
    check("insert_row has 'container'",             "container" in params)
    check("insert_row has 'position'",         "position" in params)
    check("insert_row has 'position_index'",   "position_index" in params)

    sig = inspect.signature(W.remove_element)
    params = list(sig.parameters.keys())
    check("remove_element has 'view_id'",  "view_id" in params)


def test_insert_element_unknown_uuid():
    """insert_element on an unknown UUID must raise RuntimeError."""
    print("\n=== insert_element: unknown UUID raises RuntimeError ===")

    # C layer
    try:
        _actionui.insert_element(_FAKE_UUID, 1, _VALID_ELEMENT_JSON)
        check("_actionui.insert_element(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("_actionui.insert_element(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"_actionui.insert_element raised unexpected {type(e).__name__}: {e}", False)

    # Python wrapper
    win = actionui.Window(_FAKE_UUID)
    try:
        win.insert_element(1, _VALID_ELEMENT_JSON)
        check("Window.insert_element(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("Window.insert_element(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"Window.insert_element raised unexpected {type(e).__name__}: {e}", False)


def test_insert_element_dict_input():
    """Window.insert_element must accept a dict and convert it to JSON."""
    print("\n=== insert_element: dict input coerced to JSON ===")

    win = actionui.Window(_FAKE_UUID)
    element_dict = {"type": "Text", "properties": {"text": "hello"}}

    try:
        win.insert_element(1, element_dict)
        # On unknown UUID this raises RuntimeError — the dict path still ran
        check("insert_element(dict) coerced to JSON (RuntimeError from unknown UUID)", False)
    except RuntimeError:
        check("insert_element(dict) coerced to JSON (RuntimeError from unknown UUID)", True)
    except TypeError as e:
        check(f"insert_element(dict) raised TypeError (dict not serialised): {e}", False)


def test_insert_row_unknown_uuid():
    """insert_row on an unknown UUID must raise RuntimeError."""
    print("\n=== insert_row: unknown UUID raises RuntimeError ===")

    # C layer
    try:
        _actionui.insert_row(_FAKE_UUID, 2, _VALID_CELLS_JSON)
        check("_actionui.insert_row(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("_actionui.insert_row(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"_actionui.insert_row raised unexpected {type(e).__name__}: {e}", False)

    # Python wrapper
    win = actionui.Window(_FAKE_UUID)
    try:
        win.insert_row(2, _VALID_CELLS_JSON)
        check("Window.insert_row(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("Window.insert_row(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"Window.insert_row raised unexpected {type(e).__name__}: {e}", False)


def test_insert_row_list_input():
    """Window.insert_row must accept a list of dicts and convert to JSON."""
    print("\n=== insert_row: list input coerced to JSON ===")

    win = actionui.Window(_FAKE_UUID)
    cells = [
        {"type": "Text", "properties": {"text": "C0"}},
        {"type": "Text", "properties": {"text": "C1"}},
    ]

    try:
        win.insert_row(2, cells)
        check("insert_row(list) coerced to JSON (RuntimeError from unknown UUID)", False)
    except RuntimeError:
        check("insert_row(list) coerced to JSON (RuntimeError from unknown UUID)", True)
    except TypeError as e:
        check(f"insert_row(list) raised TypeError (list not serialised): {e}", False)


def test_remove_element_unknown_uuid():
    """remove_element on an unknown UUID must raise RuntimeError."""
    print("\n=== remove_element: unknown UUID raises RuntimeError ===")

    # C layer
    try:
        _actionui.remove_element(_FAKE_UUID, 10)
        check("_actionui.remove_element(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("_actionui.remove_element(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"_actionui.remove_element raised unexpected {type(e).__name__}: {e}", False)

    # Python wrapper
    win = actionui.Window(_FAKE_UUID)
    try:
        win.remove_element(10)
        check("Window.remove_element(unknown UUID) raised RuntimeError", False)
    except RuntimeError:
        check("Window.remove_element(unknown UUID) raises RuntimeError", True)
    except Exception as e:
        check(f"Window.remove_element raised unexpected {type(e).__name__}: {e}", False)


def test_position_defaults():
    """Position defaults (append=0) must flow through without TypeError."""
    print("\n=== insert_element: optional args default to 0/None ===")

    win = actionui.Window(_FAKE_UUID)

    # All optional args omitted — must not raise TypeError, only RuntimeError (unknown UUID)
    for call_desc, kwargs in [
        ("no container/position",          {}),
        ("container=None",                 {"container": None}),
        ("container='children'",           {"container": "children"}),
        ("position=PREPEND",      {"position": actionui.InsertPosition.PREPEND}),
        ("position=AT, param=0",  {"position": actionui.InsertPosition.AT, "position_param": 0}),
        ("position=BEFORE, param=9", {"position": actionui.InsertPosition.BEFORE, "position_param": 9}),
        ("position=AFTER, param=9",  {"position": actionui.InsertPosition.AFTER, "position_param": 9}),
    ]:
        try:
            win.insert_element(1, _VALID_ELEMENT_JSON, **kwargs)
            check(f"insert_element({call_desc}) raised RuntimeError", False)
        except RuntimeError:
            check(f"insert_element({call_desc}) raises RuntimeError (not TypeError)", True)
        except TypeError as e:
            check(f"insert_element({call_desc}) raised TypeError (bad default): {e}", False)


def test_insert_row_position_defaults():
    """insert_row position defaults must flow through without TypeError."""
    print("\n=== insert_row: optional args default to 0/None ===")

    win = actionui.Window(_FAKE_UUID)

    for call_desc, kwargs in [
        ("no container/position",       {}),
        ("position=PREPEND",   {"position": actionui.InsertPosition.PREPEND}),
        ("position=AT, index=0",    {"position": actionui.InsertPosition.AT, "position_index": 0}),
    ]:
        try:
            win.insert_row(2, _VALID_CELLS_JSON, **kwargs)
            check(f"insert_row({call_desc}) raised RuntimeError", False)
        except RuntimeError:
            check(f"insert_row({call_desc}) raises RuntimeError (not TypeError)", True)
        except TypeError as e:
            check(f"insert_row({call_desc}) raised TypeError (bad default): {e}", False)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    print("ActionUI Structural Mutation API Smoke Tests")
    print("=" * 55)
    print("(No NSApplication run loop — tests the C extension binding layer only)")

    # Ensure an Application instance exists (required by Window constructor path)
    if actionui.Application.instance() is None:
        actionui.Application()

    test_c_api_surface()
    test_window_method_surface()
    test_insert_element_signatures()
    test_insert_element_unknown_uuid()
    test_insert_element_dict_input()
    test_insert_row_unknown_uuid()
    test_insert_row_list_input()
    test_remove_element_unknown_uuid()
    test_position_defaults()
    test_insert_row_position_defaults()

    print("\n" + "=" * 55)
    if not _failures:
        print("All checks PASSED.")
        sys.exit(0)
    else:
        print(f"FAILED — {len(_failures)} check(s):")
        for f in _failures:
            print(f"  - {f}")
        sys.exit(1)
